### DISCUSSION TITLE
How to serve different versions of a file from the public directory?

### DISCUSSION THREAD

**Discussion by james9823** [vercel/next.js#12834] [Help] [answered]
I have a single nextjs app serving multiple domains.
I want to be able to serve different versions of /sitemap.xml depending on which domain the request is from.

I have the actual sitemap files in the public dir, like public/sitemapCOM.xml, public/sitemapDE.xml etc

I've tried to do this by adding the following to express in server.js 

  server.get("/sitemap.xml", (req, res) => {
    const domain = req.headers["x-forwarded-host"].split(".").slice(-1)[0].toUpperCase()
    if(domain === "COM") {
       app.render(req, res, "/sitemapCOM.xml");
    } else if (domain === "DE") {
       app.render(req, res, "/sitemapDE.xml");
    }
  })

I end up with getting a missing page (404)

Is there any other way of accomplishing this? Any help would be greatly appreciated

**Comment by brunocrosier** [comment]
If you're using [API routes](https://nextjs.org/docs/api-routes/api-middlewares) then Express.js stuff won't work. 

If your site is hosted on Vercel, you could look into adding Rewrites in your now.json file, maybe that would help ?

https://vercel.com/docs/configuration#project/rewrites

**Comment by leerob** [comment, marked as answer]
You could build the sitemap at `sitemap.xml.js` and use `getServerSideProps`. 

```js
export async function getServerSideProps({ req, res }) {
  const sitemapCom = createSitemap();
  const sitemapDe = createSitemap();
  const sitemap = domain === 'COM' ? sitemapCom : sitemapDe;

  res.setHeader('Content-Type', 'text/xml');
  res.write(sitemap);
  res.end();

  return {
    props: {},
  }
}
```

**Reply by brunocrosier** [reply]
This is nice!

**Reply by james9823** [reply]
This looks like a good solution! Thanks a lot!
